include/big_core_base.h

```diff
@@ -45,78 +54,105 @@ enum class DataTypes : uint8_t // RV: v big.pdf je jasne definovane, ze je to 8
     void clear();
 
     // Returns true if all data are in memory.
-    bool isInMemory();
+    bool isInMemory() { return _data != nullptr && memorySize == dataSize; }
 
     // Sets maximum possible size of memory that can be used. Does not free memory if already allocated.
-    // RV: size_t -> uint64_t viz vyse. const jsem smazal, protoze je to u beznych parametru (tedy ne pointer nebo reference) zbytecne a nepouziva se to.
-    void setMemorySize(uint64_t bytes = DEFAULT_MEMORY_SIZE);
+    void setMaxMemorySize(uint64_t bytes = DEFAULT_MEMORY_SIZE) { maxMemorySize = bytes; }
 
     // Returns total size of the data.
-    uint64_t size();
+    uint64_t size() { return dataSize; }
 
     // Returns size of the used memory.
-    uint64_t sizeInMemory();
+    uint64_t sizeInMemory() { return memorySize; }
+
+protected:
+
+    // Initializes supporting data structures.
+    void setSupportingStructures();
 
-    // returns size of givne data type
-    // RV: jmeno teto metody mi nejak nesedi s tim, co ma delat.
-    size_t getImageType(const uint64_t dataType);
+    // Do not call directly, use setSupportingStructures() method.
+    void setDimensions();
 
-    // returns width * height * planes * tiles no matter data type
-    // RV: je potreba pracovat spise s jednotlivymi dlazdicemi (tiles)
-    // RV: metoda se mi zda zbytecna, vypocet muze byt soucasti imageSize
-    size_t imageSizeRaw();
+    // Do not call directly, use setSupportingStructures() method.
+    void setEntityTypeSizes();
 
-    // returns memory size of one image
-    size_t imageSize(const uint64_t dataType);
+    // Do not call directly, use setSupportingStructures() method.
+    void setSubSizes();
 
-    // Returns true if all data have the same data type.
-    bool isUniformDataType();
+    // Do not call directly, use setSupportingStructures() method.
+    void setDataSize();
+
+    // Do not call directly, use setSupportingStructures() method.
+    void setOffsets();
+
+    // Do not call directly, use setSupportingStructures() method.
+    void setOrderMap();
 
 protected:
+
     uint64_t numberOfImages = 0;
-    uint64_t numberOfTiles = 1;
     uint64_t imageHeight = 0;
     uint64_t imageWidth = 0;
     uint64_t numberOfPlanes = 1;
-    std::vector<uint64_t> dataOrder = { 1, 2, 3, 4, 5 };
-    std::vector<uint8_t> dataType = { 2 };
-    char *data = nullptr;
-
-    uint64_t dataLength = 0;
-    uint64_t memorySize = DEFAULT_MEMORY_SIZE;
-
-    std::vector<uint64_t> outermostEntitiesOffsets;
+    std::vector<DataOrderIds> dataOrder = { DataOrderIds::NUMBER_OF_IMAGES, DataOrderIds::IMAGE_HEIGHT, DataOrderIds::IMAGE_WIDTH, DataOrderIds::NUMBER_OF_PLANES };
+    std::vector<DataTypes> dataType = { DataTypes::FLOAT };
+    char *_data = nullptr;
+
+    uint64_t dataSize = 0;                              // size of the data according to given dimensions and data types
+    uint64_t memorySize = 0;                            // size of the used memory
```

Comment:
k cemu je memorySize, kdyz ve chvili kdy se data do pameti nevejdou, tak je vubec nenacitas?

Reply:
V budoucnu bych chtěl mít alokovaný buffer (o velikosti memorySize), do kterého se budou data číst v případě, že je po nich poptávka. Do bufferu se budou ukládat jednotlivé outermost entity. Pokud bude dataOrder = {1,*,*,*}, budou se do bufferu načítat vždy celé obrázky. Pokud například chci pixel z obrázku číslo 25, načte se celý do bufferu a vrátí se daný pixel. Pokud pak v budoucnu budu chtít zase nějaký pixel z obrázku 25, sáhne se do bufferu. Když v bufferu dojde místo, nahradí se nejdelší dobu nepoužitý obrázek. Pokud uživatel nechce buffer používat, nastaví si maxMemorySize = 0. Trochu problém bude, že různé obrázky můžou zabírat různou velikost, protože můžou mít různý datový typ. Pak bude nahrání obrázku s typem double do plného bufferu kde jsou samé obrázky typu char trochu problém, protože se bude muset zahodit 8 obrázků vedle sebe, aby se tam ten nový vešel. To si musím ještě rozmyslet. Možná by bylo nejlepší alokovat pro každý obrázek novou nezávislou paměť. Pak když už velikost alokované paměti dosáhne maxima, můžu jednoduše uvolnit 8 obrázků typu char a alokovat jeden typu double. Bude to o něco pomalejší, protože se občas bude muset alokovat nová paměť. Ale zase to nebude muset na začátku alokovat všechnu paměť, pouze to bude postupně přidávat novou paměť.